librosa/sequence.py

```diff
@@ -359,6 +361,341 @@ def __dtw_backtracking(D_steps, step_sizes_sigma):  # pragma: no cover
+    ...                                        knight_moves=False)
+    >>> plt.subplot(1,2,1)
+    >>> librosa.display.specshow(rec, x_axis='frames', y_axis='frames')
+    >>> plt.title('Recurrence matrix')
+    >>> plt.subplot(1,2,2)
+    >>> librosa.display.specshow(L_score, x_axis='frames', y_axis='frames')
+    >>> plt.plot(L_path[:, 1], L_path[:, 0], label='Optimal path', color='c')
+    >>> plt.legend()
+    >>> plt.show()
+
+    Full alignment using gaps and knight moves
+
+    >>> # New gaps cost 5, extending old gaps cost 10 for each step
+    >>> score, path = librosa.sequence.rqa(rec, 5, 10)
+    >>> plt.figure()
+    >>> plt.subplot(1,2,1)
+    >>> librosa.display.specshow(rec, x_axis='frames', y_axis='frames')
+    >>> plt.title('Recurrence matrix')
+    >>> plt.subplot(1,2,2)
+    >>> librosa.display.specshow(score, x_axis='frames', y_axis='frames')
+    >>> plt.plot(path[:, 1], path[:, 0], label='Optimal path', color='c')
+    >>> plt.legend()
+    >>> plt.show()
+    '''
+
+    if gap_onset < 0:
+        raise ParameterError('gap_onset={} must be strictly positive')
+    if gap_extend < 0:
+        raise ParameterError('gap_extend={} must be strictly positive')
+
+    score, pointers = __rqa_dp(sim, gap_onset, gap_extend, knight_moves)
+    if backtrack:
+        path = __rqa_backtrack(score, pointers)
+        return score, path
+
+    return score
+
+
+@jit(nopython=True, cache=True)
+def __rqa_dp(sim, gap_onset, gap_extend, knight):  # pragma: no cover
+    '''RQA dynamic programming implementation'''
+
+    # The output array
+    score = np.zeros(sim.shape, dtype=sim.dtype)
+
+    # The backtracking array
+    backtrack = np.zeros(sim.shape, dtype=np.int8)
+
+    # These are place-holder arrays to limit the points being considered
+    # at each step of the DP
+    #
+    # If knight moves are enabled, values are indexed according to
+    # [(-1,-1), (-1, -2), (-2, -1)]
+    #
+    # If knight moves are disabled, then only the first entry is used.
+    #
+    # Using dummy vectors here makes the code a bit cleaner down below.
+    #
+    sim_values = np.zeros(3)
+    score_values = np.zeros(3)
+    vec = np.zeros(3)
+
+    if knight:
+        # Initial limit is for the base case: diagonal + one knight
+        init_limit = 2
+
+        # Otherwise, we have 3 positions
+        limit = 3
+    else:
+        init_limit = 1
+        limit = 1
+
+    # backtracking rubric:
+    #   0 ==> diagonal move
+    #   1 ==> knight move up
+    #   2 ==> knight move left
+    #  -1 ==> reset without inclusion
+    #  -2 ==> reset with inclusion (ie positive value at init)
+
+    # Initialize the first row and column with the data
+    score[0, :] = sim[0, :]
+    score[:, 0] = sim[:, 0]
+
+    # backtracking initialization: the first row and column are all resets
+    # if there's a positive link here, it's an inclusive reset
+    for i in range(sim.shape[0]):
+        if sim[i, 0]:
```

Comment:
⚠️ ⚠️ ⚠️ 
why is it `if sim[i,0]` and not `if sim[i,0]>0`? can `sim` contain `None` or `False`?

Reply:
sim can't contain none.  It could contain `False` (dtyped to np.bool); in general, nonzero is truthy and zero is not truthy.

Reply:
ok i understand now!